p4-16/psa/PSA.mdk

```diff
@@ -199,6 +202,41 @@ copy going to the control CPU to have a software-defined header
 (defined in the P4 program) that is different than any headers on the
 packet copies going to the Egress control block.
 
+A PSA implementation may choose to implement multiple class-of-service
+queues for each output port, but it may also choose to implement only
+1 queue for each output port.  The Ingress control block may choose to
+assign a value to the `ostd.class_of_service` field to change the
+packet's class of service to a value other than the default of 0.
+
+PSA only specifies how the Ingress control block can choose among
+multiple class-of-service queues for an output port, if more than one
+exists.  It does not mandate a scheduling policy among these queues.
+Something at least as flexible as weighted fair queuing, with an
```

Comment:
I would remove this sentence. 

I would add as a note, though, that queueing and scheduling have the potential (if not outright intended) to reorder packets. This can have visible effect on the P4 programs, especially those containing stateful elements, writable by the data plane (counters, meters, registers). Any tests written for such systems must account for such differences in behavior otherwise they will not be portable.

Reply:
If it isn't outright intended to reorder packets, then there isn't much point to having multiple classes of service, is there?  The only reason to introduce them that I can think of _other_ than to explicitly allow reordering between different classes of service, is if the class of service affected some drop policy in the PRE.

Reply:
I think we agree, that packets will have a god chance of being reordered.

PSA allows users to write programs that are sensitive to packet ordering (mostly via registers). Therefore, it might make sense to warn people, but, I might be wrong on that point.

Reply:
Actually, after waking up in the morning I realized that even with a single queue the order of the packet on the egress can still be different (e.g. when packets from multiple ports go to the same port you never know the exact order), so, perhaps, this is so fundamental that no additional warnings are necessary.

Reply:
The sentence about recommended weighted fair queueing scheduling policy is still in there, but note that it is only a recommendation, not mandated.